商品出品機能

Tsutomu-sasaki1 opened the pull request:
# What
商品出品機能

#Why
商品出品機能を実装するため

#単体テスト画面
https://gyazo.com/ad75752005766df4243e8ed3776eb34d

#ログイン状態のユーザーだけが、商品出品ページへ遷移できる
https://gyazo.com/a9e028e8b9de97444e894fa75b69f23b

#ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する
https://gyazo.com/19df88a049d3d92d3ab0c99e19153892

#入力された販売価格によって、販売手数料や販売利益の表示が変わる
https://gyazo.com/998fae0fd87380b4004d05a66a6608f0

#必要な情報を適切に入力すると、商品の出品ができる
https://gyazo.com/3b268905122c5a6ade8f5af9a592f95e
https://gyazo.com/87dc32498d0fc1c2050628f6c58bf08d

#入力に問題がある状態で出品ボタンが押されたら、出品ページに戻りエラーメッセージが表示されること
（エラーハンドリングができている）
https://gyazo.com/1c292f3cc9fe6391eb9fbe28a5baaeef

よろしくお願いします。

＃修正後のテスト挙動
https://gyazo.com/be96436fc8020729f34ea860e7dbd646

よろしくお願いします。

＃再々修正　テスト挙動
https://gyazo.com/7761fa7d5d81de7b88b3c3cf884c8387

よろしくお願いします。

#再々々修正　テスト挙動
https://gyazo.com/6c3dd7550749bfbf52db656476a913d4

よろしくお願いします。

Tsutomu-sasaki1 committed:
itemモデル、マイグレーション　設定

Tsutomu-sasaki1 committed:
Active Storageの導入

Tsutomu-sasaki1 committed:
ActiveHash導入~アソシエーション

Tsutomu-sasaki1 committed:
new,createアクション実装

Tsutomu-sasaki1 committed:
validates　inclusion実装

Tsutomu-sasaki1 committed:
単体テスト完了

Tsutomu-sasaki1 committed:
Itemクラス　バリデーション追加

maimai0401 commented on app/models/item.rb:
```diff
@@ -0,0 +1,18 @@
+
+  has_one :history
+  has_one_attached :image
+  belongs_to :genre, :prefecture, :condition, :dateofshipment, :shippingharge, :user
```
```
  belongs_to :genre, :prefecture, :condition, :dateofshipment, :shippingharge, :user
```
こちらですが、
```
belongs_to :genre
belongs_to :prefecture
```
のように、一つずつ記述しましょう。
現在の記述だとエラーが出る可能性があるからです。

Tsutomu-sasaki1 replied:
修正後、商品登録画面でshippinghargeの入力が無効になるエラーが出たので元に戻しました。

maimai0401 commented on app/models/item.rb:
```diff
@@ -0,0 +1,18 @@
+class Item < ApplicationRecord
+  extend ActiveHash::Associations::ActiveRecordExtensions
+
+  validates :item_name,            presence: true
```
```
validates :item_name,            presence: true
  validates :category_id,          presence: true, numericality: { other_than: 1}
  validates :condition_id,         presence: true, numericality: { other_than: 1}
  validates :prefectures_id,       presence: true, numericality: { other_than: 1}
  validates :date_of_shipment_id,  presence: true, numericality: { other_than: 1}
  validates :shipping_charge_id,   presence: true, numericality: { other_than: 1}
  validates :price,                presence: true, inclusion: { in: 300..9999999  }
  validates :explanation,          presence: true
  validates :image,                presence: true
```
with_optionsを使用してpresence: trueのバリデーションをまとめましょう。

Tsutomu-sasaki1 replied:
修正しました。

maimai0401 commented on app/models/item.rb:
```diff
@@ -0,0 +1,18 @@
+  extend ActiveHash::Associations::ActiveRecordExtensions
+
+  validates :item_name,            presence: true
+  validates :category_id,          presence: true, numericality: { other_than: 1}
```
```
validates :category_id,          presence: true, numericality: { other_than: 1}
  validates :condition_id,         presence: true, numericality: { other_than: 1}
  validates :prefectures_id,       presence: true, numericality: { other_than: 1}
  validates :date_of_shipment_id,  presence: true, numericality: { other_than: 1}
  validates :shipping_charge_id,   presence: true, numericality: { other_than: 1}
```
with_optionsを使用してnumericality: { other_than: 1}のバリデーションをまとめましょう。

Tsutomu-sasaki1 replied:
修正しました。

maimai0401 commented on app/models/list.rb:
```diff
@@ -0,0 +1,10 @@
+# class List < ApplicationRecord
```
こちらのファイルは使用しないようでしたら削除しておきましょう。

Tsutomu-sasaki1 replied:
修正しました。

maimai0401 commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,18 @@
+    prefectures_id       {2}
+    date_of_shipment_id  {2}
+    shipping_charge_id   {2}
+
```
ユーザーとのアソシエーション を追記しましょう。

Tsutomu-sasaki1 replied:
修正しました。

maimai0401 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,79 @@
+      it "商品画像を１枚つける、商品名、説明を入力、配送料、カテゴリー、商品状態、発送元地域、発送までの日数の選択、販売価格を半角英数字で入力。以上を満たせば登録可能" do
+        expect(@item).to be_valid
+      end
+    end
```
より正確なテストを行うために、下記テストコードを実装しましょう。
・販売価格は全角文字では登録できないこと
・販売価格は半角英数混合では登録できないこと
・販売価格は半角英語だけでは登録できないこと

Tsutomu-sasaki1 replied:
修正しました。

maimai0401 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,79 @@
+        expect(@item.errors.full_messages).to include("Price can't be blank")
+      end
+      
+      it "価格の範囲が、¥300~¥9,999,999の間であること" do
```
'価格の範囲が¥300~¥9,999,999の間ではない場合、商品の出品ができない'
こちらのテストですが、境界値をテストする事が一般的です。 そのため、299円の場合と10,000,000円の場合の2パターンをテストしましょう。

Tsutomu-sasaki1 replied:
修正しました。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
7点コメントいたしましたので、ご確認をお願いいたします。

Tsutomu-sasaki1 committed:
バリデーション修正、Listモデル削除

Tsutomu-sasaki1 committed:
factorybot 修正

Tsutomu-sasaki1 committed:
テストコード　追加

tatsuki-kato commented on app/models/item.rb:
```diff
@@ -0,0 +1,27 @@
+
+  has_one :history
+  has_one_attached :image
+  belongs_to :genre, :prefecture, :condition, :dateofshipment, :shippingharge, :user
```
```ruby
belongs_to :genre, :prefecture, :condition, :dateofshipment, :shippingharge, :user
```
こちらの記述ですが、前回のレビュワーからコメントがあったように、
```
belongs_to :genre
belongs_to :prefecture
```
のように、一つずつ記述しましょう。
現在の記述だとエラーが出る可能性があるからです。

もし、修正後再度エラーが出るようであれば、一度通話メンターへご質問いただけると幸いです。

Tsutomu-sasaki1 replied:
修正しました。

tatsuki-kato commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,41 @@
+    @item = Item.new
+  end
+
+  def create
+    @item = Item.create(item_params)
+    if @item.save
```
```ruby
 @item = Item.create(item_params)
```
こちらは`@item = Item.new(item_params)`にしましょう。
createメソッドを使うとエラーハンドリングする前に情報が保存されてしまうためです。

Tsutomu-sasaki1 replied:
修正しました。

tatsuki-kato commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,18 @@
+    date_of_shipment_id  {2}
+    shipping_charge_id   {2}
+
+    user_id = 1   #userとのアソシエーション
```
```ruby
user_id = 1   #userとのアソシエーション
```
こちら`association :user`と記述しましょう。
この記述をすることで、FactoryBotでitemのインスタンスを生成すると同時にuserのインスタンスも生成することができるためです。

Tsutomu-sasaki1 replied:
修正しました。

tatsuki-kato commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,102 @@
+      end
+      
+      it "価格の範囲が、¥300~¥9,999,999の間であること" do
+        @item.price = "299"
```
```ruby
@item.price = "299"
```
こちらダブルクォーテーションは外しましょう。
priceカラムは数値型であるためです。

同様に68行目の
```ruby
@item.price = "10000000"
```
もダブルクォーテーションを外しましょう。

Tsutomu-sasaki1 replied:
修正しました。

tatsuki-kato reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、確認をお願いします！

Tsutomu-sasaki1 committed:
createアクション　修正

Tsutomu-sasaki1 committed:
itemテストコード修正(アソシエーション)

Tsutomu-sasaki1 committed:
Itemモデル　アソシエーション修正

Tsutomu-sasaki1 committed:
単体テスト　最終確認

sgwr-0113 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,103 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Explanation can't be blank")
+       end
+      it "カテゴリーの情報が必須であること" do
+        @item.category_id = nil
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Category can't be blank", "Category is not a number")
+      end
+      
+      it "商品の状態についての情報が必須であること" do
+        @item.condition_id = nil
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Condition can't be blank", "Condition is not a number")
+      end
+      
+      it "発送元の地域についての情報が必須であること" do
+        @item.prefectures_id = nil
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Prefectures can't be blank", "Prefectures is not a number")
+      end
+      
+      it "配送料の負担についての情報が必須であること" do
+        @item.shipping_charge_id = nil
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Shipping charge can't be blank", "Shipping charge is not a number")
+      end
+      
+      it "発送までの日数についての情報が必須であること" do
+        @item.date_of_shipment_id = nil
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Date of shipment can't be blank", "Date of shipment is not a number")
+      end
```
26~54行目
「必須であること」だけではなく、**「1以上であること」**（何かしらが選択されていること）についてもそれぞれテストをしましょう。
（`numericality: { other_than: 1}`のバリデーションを記述しているため）

Tsutomu-sasaki1 replied:
修正しました。

Tsutomu-sasaki1 replied:
修正しました。

sgwr-0113 commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,18 @@
+FactoryBot.define do
+  factory :item do
+    id                   {Faker::Number.unique.number(1)}
+    item_name            {Faker::Commerce}
+    explanation          {Faker::Lorem}
+    price                {Faker::Number.between(from: 300, to: 9999999)}
+    category_id          {Faker::Number.number(2..10)}
+    condition_id         {Faker::Number.number(2..7)}
+    prefectures_id       {Faker::Number.number(2..47)}
+    date_of_shipment_id  {Faker::Number.number(2..4)}
+    shipping_charge_id   {Faker::Number.number(2..3)}
```
3~11行目

こちらの表現でもテストは通りますが、Fakerは**一意性制約**があるものに使うのが一般的です。
例えば、ユーザーのemailにはFakerを使ってランダム性を与えないと、2人目の生成ができなくなってしまいますよね。

Factoryでは**とりあえず登録できる情報**さえ設定できていればいいので、一意性制約が特にない場合はあまり難しく考えず、

```rb
item_name              { "サンプル商品" }
explanation             { "サンプル商品の説明" }
price             { 1000 }
category_id 　       { 2 }

```
のように、具体的な値を入れると見栄えの良い記述になります。

Tsutomu-sasaki1 replied:
修正しました。

sgwr-0113 commented on db/migrate/20210203054856_create_prefectures.rb:
```diff
@@ -0,0 +1,8 @@
+class CreatePrefectures < ActiveRecord::Migration[6.0]
+  def change
+    create_table :prefectures do |t|
+
+      t.timestamps
+    end
+  end
+end
```
こちらのマイグレーションファイルは不要ですので、**rollbackしてから削除しておきましょう。**

Tsutomu-sasaki1 replied:
修正しました。

sgwr-0113 commented on db/migrate/20210203055015_create_shippingcharges.rb:
```diff
@@ -0,0 +1,8 @@
+class CreateShippingcharges < ActiveRecord::Migration[6.0]
+  def change
+    create_table :shippingcharges do |t|
+
+      t.timestamps
+    end
+  end
+end
```
こちらのマイグレーションファイルは不要ですので、**rollbackしてから削除しておきましょう。**

Tsutomu-sasaki1 replied:
修正しました。

sgwr-0113 commented on db/migrate/20210206092005_me.rb:
```diff
@@ -0,0 +1,5 @@
+class Me < ActiveRecord::Migration[6.0]
+  def change
+    drop_table :shippingcharges
+  end
+end
```
こちらのマイグレーションファイルは不要ですので、**rollbackしてから削除しておきましょう。**

Tsutomu-sasaki1 replied:
修正しました。

sgwr-0113 commented on db/migrate/20210206092636_create_dateofshipments.rb:
```diff
@@ -0,0 +1,8 @@
+class CreateDateofshipments < ActiveRecord::Migration[6.0]
+  def change
+    create_table :dateofshipments do |t|
+
+      t.timestamps
+    end
+  end
+end
```
こちらのマイグレーションファイルは不要ですので、**rollbackしてから削除しておきましょう。**

Tsutomu-sasaki1 replied:
修正しました。

sgwr-0113 reviewed (commented):
レビュー依頼ありがとうございます！
追加でコメントしました。
やや細かい点で恐縮ですが、正しくレビューを行うためにも、修正をお願いできればと存じます。

修正後、再度フォームにてレビュー依頼をお願いいたします。

Tsutomu-sasaki1 committed:
レビュー　修正

各修正毎のコミット忘れていました。
以下修正点です
・不要なマイグレーションファイル削除
・factorybotの内容修正
・プルダウンメニューのテストコード追加

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認いたしました。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。
